### DISCUSSION TITLE
Migrating from LND while preserving seed

### DISCUSSION THREAD

**Discussion by kiwiidb** [ACINQ/eclair#2581] [Q&A]
Would it be possible to migrate from an LND node and keep the same public key? I'm not talking about channels or even recovering on-chain funds - just the same node id. Is it possible to go  from LND's 24 words to a `node_seed.dat`  usable by Eclair?

**Comment by t-bast** [comment]
That's a good question, I know that lnd has been using their own derivation for keys, so I'm not sure this is possible. Eclair's `node_seed.dat` is just random bytes used to create a standard BIP32 xpriv, but if lnd doesn't use BIP32 derivation, that won't yield the same `nodeId`...

**Reply by kiwiidb** [reply]
What about this: https://github.com/guggero/chantools#seed-and-passphrase-input

**Reply by t-bast** [reply]
I have no idea, I don't know anything about lnd's internal key derivation.
But you can probably test this and see if you're able to come up with the same `nodeId`?
Eclair's `node_seed.dat` and `channel_seed.dat` are just binary files that contain the seed bytes.

**Reply by kiwiidb** [reply]
Will test & report back.

**Reply by kiwiidb** [reply]
I've been testing out a few things.
- Use chantools to dump LND's [cipherSeed.Entropy](https://github.com/guggero/chantools/blob/master/lnd/aezeed.go#L114). Notice that the seed is only 16 bytes instead of 32, but I am not sure if that actually matters here.
- Figure out the default derivation paths used by LND / Eclair. For LND (regtest) I think that the derivation path is something like `m/1017'/1'/6'/0` ([purpose](https://github.com/lightningnetwork/lnd/blob/master/keychain/derivation.go#L35)/ [account](https://github.com/lightningnetwork/lnd/blob/master/keychain/derivation.go#L103))
- So I've been trying to very [crudely change](https://github.com/getAlby/eclair/pull/1/files) what I think is Eclair's derivation path for the node key to match LND's. Then I copy over the 16-byte `node_seed.dat` that I dumped from the `cipherSeed` mentioned above and start Eclair.
Alas, no luck, the pubkey is still not the same :/

**Reply by t-bast** [reply]
> Notice that the seed is only 16 bytes instead of 32, but I am not sure if that actually matters here.

This shouldn't matter, we simply feed this seed into a standard BIP 32 HD wallet, so as long as lnd does the same, it should work with only 16 bytes.

> I've been trying to very [crudely change](https://github.com/getAlby/eclair/pull/1/files) what I think is Eclair's derivation path for the node key

The one you should change is here: https://github.com/ACINQ/eclair/blob/master/eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalNodeKeyManager.scala

The one you changed is for channel keys (keys used for channel transactions).

Eclair explicitly separates the `node_seed` (keys used for encryption tcp connections and responsible for your `node_id`) from `channel_seed` (keys used for on-chain transactions). These two types of keys don't have the same liveness requirements, that's why we separated them.

Hopefully that will make it work?
